Debug Menu extensions: Display strings & interpreter stacks

florianessl opened the pull request:
Adds two new items to the F9 Debug Menu:
- "Strings"
- "Interpreter"

String display supports viewing multiline strings through a special new window "Window_StringView"
Two options can be toggled:
- Automatic linebreaks
- Command Evaluation ( default inserter with \v, \n, \t )

![screenshot_160](https://github.com/EasyRPG/Player/assets/42896817/b3821b19-7671-4d83-a967-03da85422c86)

Viewing the current interpreter stacks
(right side: [current command] / [commands_size])
![screenshot_161](https://github.com/EasyRPG/Player/assets/42896817/2c01a762-e96e-4659-9fea-c1bce15632ad)

Ghabry commented:
For the Linux build you must update the Makefile.am file similar to CMakeLists

Ghabry commented:
Haven't read the entire code yet obviously. Just one thing I noticed: 

You wrote your own wrap code. We already have a word wrapping function Game_Message::WordWrap. Maybe use this one if applicable?

florianessl commented:
> Haven't read the entire code yet obviously. Just one thing I noticed:
> 
> You wrote your own wrap code. We already have a word wrapping function Game_Message::WordWrap. Maybe use this one if applicable?

I switched the code, so it uses this WordWrapping function now, instead of just cutting of at the exact position a text would overflow.

florianessl commented:
Rebased & refactored a bit. The ScopedVars branch depends on this, to make it possible to view & edit all frame-local variables for individual interpreter frames. I already moved some of the new code here (New choices window for debug scene) to make things easier to merge.

Considering the necessary changes for the debugging possibilities of the ScopedVars branch, this extension should be feature-complete, so feel free to test this!

Ghabry commented:
Cool. We do not have too many games that use String Vars online.

This test game here could be suitable: https://easyrpg.org/play/pr3219/?game=string-var&test-play

I think I found a bug: In this game when I go to Interpreter -> Main no cursor appears.

However when I try e.g. uid https://easyrpg.org/play/pr3219/?game=uid&test-play  it works. So I guess only having one interpreter bugs the Ui here?

Ghabry commented:
The issue is still not completely fixed: When you go in the string-var game to Strings -> Main the cursor is broken the first time. After Cancel -> Decision again it works :shrug:

Ghabry commented on src/window_interpreter.cpp:
```diff
@@ -0,0 +1,181 @@
+				} else if (com.parameters[0] == 3 && Player::IsPatchManiac()) {
+					evt_id = Main_Data::game_variables->Get(com.parameters[1]);
+				} else if (com.parameters[0] == 4 && Player::IsPatchManiac()) {
+					evt_id = Main_Data::game_variables->GetIndirect(com.parameters[1]);
```
Are you attempting to figure out the event ID here?

Actually Maniac Patch added some debug fields to the ExexFrame for this purpose: ``maniac_event_info``, ``maniac_event_id``, ``maniac_event_page_id``. We do not have this implemented and you do not have to implement everything but ``maniac_event_id`` could be useful to track the ID across maps because its not reset to 0.

florianessl replied:
Yeah, that's the idea..., I've noticed that there are several unimplemented fields, but skipped over them because I don't have any implementation details and didn't want to invest the time right now to figure this out. 😬
So this is far from perfect right now.
But yeah, I should probably change the tracked id field there. Completely forgot that the vanilla event_id gets reset.
I'm on holiday right now and won't have much time to look over this. If you want to change something, feel free to do so.

Ghabry replied:
No Problem. Is nothing urgent Is enough when you annotate it with a fixme for now.

Ghabry reviewed (commented):
My bar for Debug Scene changes is quite low as it cannot break code and is - well - for debugging. Imo that code can be merged as is after the two issues are discussed/resolved.

florianessl commented:
I added some "FIXME" comments for stuff that might need some minor rework, fixed some minor display issues & added support for the maniac fields "maniac_event_id" & "maniac_event_page_id".
I couldn't reproduce any more cursor issues with the strings example game though

Here's some example of a parallel process map event calling a CE, calling another map event, calling another CE...
![screenshot_165](https://github.com/EasyRPG/Player/assets/42896817/f3c079bb-b23d-4cf1-8ed4-ff03126f74c6)
(Had to highlight this one, because this is some really unsafe stuff were it might be useful for the engine to warn the developer if such bounds are crossed -> on my list...)

Ghabry reviewed (approved):
Besides this one nit: LGTM :+1:

florianessl commented on src/window_interpreter.cpp:
```diff
@@ -54,12 +54,21 @@ void Window_Interpreter::SetStackState(bool is_ce, std::string interpreter_desc,
 	for (int i = state.stack.size() - 1; i >= 0; i--) {
 		int evt_id = state.stack[i].event_id;
+		int page_id = 0;
+		if ((Player::IsPatchManiac() || Player::HasEasyRpgExtensions()) && state.stack[i].maniac_event_id > 0) {
```
Yup, guessed the same but I stuck with the strict approach. :D 
-> Will fix
